File: AOM/packages/aom-protocol-ts/src/analyzer.ts
import type { EvidenceRef } from "./aom.js";
import type {
  RawAction,
  RawActionResult,
  RawEvent,
  RawRuntimeSnapshot,
  RawStaticSnapshot,
} from "./raw.js";
import type { TargetDescriptor } from "./target.js";

export interface AnalyzerSessionConfig {
  target: TargetDescriptor;
  artifactLocator?: string;
  executablePath?: string;
  adapterId?: string;
}

export type AnalyzerCommand =
  | { commandType: "initialize"; data: AnalyzerSessionConfig }
  | { commandType: "collect_static" }
  | { commandType: "collect_runtime" }
  | { commandType: "drain_events" }
  | { commandType: "execute_action"; data: RawAction }
  | { commandType: "shutdown" };

export interface AnalyzerReady {
  adapterId?: string;
  probeId?: string;
  evidence: EvidenceRef[];
}

export interface AnalyzerResult<T> {
  value: T;
  evidence: EvidenceRef[];
}

export interface AnalyzerFailure {
  code: string;
  message: string;
  retryable: boolean;
  evidence: EvidenceRef[];
}

export type AnalyzerReply =
  | { replyType: "ready"; data: AnalyzerReady }
  | { replyType: "static_snapshot"; data: AnalyzerResult<RawStaticSnapshot> }
  | { replyType: "runtime_snapshot"; data: AnalyzerResult<RawRuntimeSnapshot> }
  | { replyType: "events"; data: AnalyzerResult<RawEvent[]> }
  | { replyType: "action_result"; data: AnalyzerResult<RawActionResult> }
  | { replyType: "ack"; data: AnalyzerResult<boolean> }
  | { replyType: "error"; data: AnalyzerFailure };

File: AOM/crates/aom-protocol-rs/src/analyzer.rs
use crate::{
    EvidenceRef, RawAction, RawActionResult, RawEvent, RawRuntimeSnapshot, RawStaticSnapshot,
    TargetDescriptor,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AnalyzerSessionConfig {
    pub target: TargetDescriptor,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub artifact_locator: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub executable_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub adapter_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case", tag = "commandType", content = "data")]
pub enum AnalyzerCommand {
    Initialize(AnalyzerSessionConfig),
    CollectStatic,
    CollectRuntime,
    DrainEvents,
    ExecuteAction(RawAction),
    Shutdown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AnalyzerReady {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub adapter_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub probe_id: Option<String>,
    #[serde(default)]
    pub evidence: Vec<EvidenceRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AnalyzerResult<T> {
    pub value: T,
    #[serde(default)]
    pub evidence: Vec<EvidenceRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AnalyzerFailure {
    pub code: String,
    pub message: String,
    #[serde(default)]
    pub retryable: bool,
    #[serde(default)]
    pub evidence: Vec<EvidenceRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case", tag = "replyType", content = "data")]
pub enum AnalyzerReply {
    Ready(AnalyzerReady),
    StaticSnapshot(AnalyzerResult<RawStaticSnapshot>),
    RuntimeSnapshot(AnalyzerResult<RawRuntimeSnapshot>),
    Events(AnalyzerResult<Vec<RawEvent>>),
    ActionResult(AnalyzerResult<RawActionResult>),
    Ack(AnalyzerResult<bool>),
    Error(AnalyzerFailure),
}
